src/control/tools/BaseStrokeHandler.cpp

```diff
@@ -178,3 +181,60 @@ void BaseStrokeHandler::onButtonPressEvent(const PositionInputData& pos)
 		createStroke(Point(x, y));
 	}
 }
+
+
+void BaseStrokeHandler::setModifiers(double width, double height, const PositionInputData& pos, bool changeCursor)
+{
+	XOJ_CHECK_TYPE(BaseStrokeHandler);
+		
+	bool gestureShift = false;
+	bool gestureControl = false;
+	
+	if( this->drawModifier == NONE){
+		gestureShift = (width  < 0);
+		gestureControl =  (height < 0 );
+		
+		double zoom = xournal->getZoom();
+		double fixate_Dir_Mods_Dist = FIX_DRAW_TYPE_LOCK_DISTANCE / zoom; 
+		if (std::abs(width) > fixate_Dir_Mods_Dist ||  std::abs(height) > fixate_Dir_Mods_Dist )
+		{
+			this->drawModifier = (DIRSET_MODIFIERS)(SET |
+				(gestureShift? SHIFT:NONE) |
+				(gestureControl? CONTROL:NONE) );
+			if(changeCursor)
+			{
+				xournal->getCursor()->updateCursor();
+			}
+		}
+		else
+		{
+			if (changeCursor)
+			{
+				switch (( gestureShift?0:1 ) + ( gestureControl?2:0))
+				{
+					case 1:
+						xournal->getCursor()->setTempCursor(GDK_BOTTOM_RIGHT_CORNER);
+						break;
+					case 3:
+						xournal->getCursor()->setTempCursor(GDK_TOP_RIGHT_CORNER);
+						break;
+					case 0:
+						xournal->getCursor()->setTempCursor(GDK_BOTTOM_LEFT_CORNER);
+						break;
+					case 2:
+						xournal->getCursor()->setTempCursor(GDK_TOP_LEFT_CORNER);
+						break;
+				}
+			}
+		}
+	}
+	else
+	{
+		gestureShift = this->drawModifier & SHIFT;
+		gestureControl = this->drawModifier & CONTROL;
+	}
+	
+	this->modShift = pos.isShiftDown() ==  !gestureShift;
+	this->modControl = pos.isControlDown() == !gestureControl;
```

Comment:
The actual shift and Control keys will toggle any directional modifier... which may be confusing?  Not sure about this. try it out and comment.  Could make this only toggle when the directional modifier has been fixed ( which happens once the user has expanded the rect/circle past a certain point).